viogpu/viogpu3d/viogpu_allocation.cpp

```diff
@@ -0,0 +1,354 @@
+
+    DbgPrint(TRACE_LEVEL_VERBOSE, ("<--- %s\n", __FUNCTION__));
+}
+
+void VioGpuAllocation::AttachBacking(MDL* pMDL, size_t pageCount, size_t pageOffset) {
+    DbgPrint(TRACE_LEVEL_VERBOSE, ("---> %s res_id=%d\n", __FUNCTION__, m_Id));
+
+    m_pMDL = pMDL;
+    m_pageCount = pageCount;
+    m_pageOffset = pageOffset;
+
+
+    GPU_MEM_ENTRY* ents = new(NonPagedPoolNx) GPU_MEM_ENTRY[pageCount];
+
+    for (UINT i = 0; i < pageCount; i++)
+    {
+        ents[i].addr = MmGetMdlPfnArray(pMDL)[pageOffset + i] * PAGE_SIZE;
+        ents[i].length = PAGE_SIZE;
+        ents[i].padding = 0;
+    }
+
+    m_adapter->ctrlQueue.AttachBacking(m_Id, ents, (UINT)pageCount);
+    DbgPrint(TRACE_LEVEL_VERBOSE, ("<--- %s\n", __FUNCTION__));
+}
+
+void VioGpuAllocation::DetachBacking() {
+    DbgPrint(TRACE_LEVEL_VERBOSE, ("---> %s res_id=%d\n", __FUNCTION__, m_Id));
+
+    m_pMDL = NULL;
+    m_pageCount = 0;
+    m_pageOffset = 0;
+
+    m_adapter->ctrlQueue.DetachBacking(m_Id);
+    DbgPrint(TRACE_LEVEL_VERBOSE, ("<--- %s\n", __FUNCTION__));
+}
+
+
+PAGED_CODE_SEG_BEGIN
+
+void VioGpuAllocation::MarkBusy() {
+    PAGED_CODE();
+
+    DbgPrint(TRACE_LEVEL_VERBOSE, ("<--> %s res_id=%d\n", __FUNCTION__, m_Id));
+
+    InterlockedIncrement(&m_busy);
+    KeClearEvent(&m_busyNotification);
+}
+
+void VioGpuAllocation::UnmarkBusy() {
+    PAGED_CODE();
+
+    DbgPrint(TRACE_LEVEL_VERBOSE, ("<--> %s res_id=%d\n", __FUNCTION__, m_Id));
+
+    if (InterlockedDecrement(&m_busy) == 0) {
+        KeSetEvent(&m_busyNotification, IO_NO_INCREMENT, FALSE);
+    }
+}
+
+
+D3DDDIFORMAT VioGpuToD3DDDIColorFormat(virtio_gpu_formats format)
+{
+    PAGED_CODE();
+
+    switch (format)
+    {
+    case VIRTIO_GPU_FORMAT_B8G8R8A8_UNORM:
+        return D3DDDIFMT_A8R8G8B8;
+    case VIRTIO_GPU_FORMAT_B8G8R8X8_UNORM:
+        return D3DDDIFMT_X8R8G8B8;
+    case VIRTIO_GPU_FORMAT_R8G8B8A8_UNORM:
+        return D3DDDIFMT_A8B8G8R8;
+    case VIRTIO_GPU_FORMAT_R8G8B8X8_UNORM:
+        return D3DDDIFMT_X8B8G8R8;
+    }
+    DbgPrint(TRACE_LEVEL_ERROR, ("---> %s Unsupported color format %d\n", __FUNCTION__, format));
+    return D3DDDIFMT_X8B8G8R8;
+}
+
+void VioGpuAllocation::FlushToScreen(UINT scan_id) {
+    PAGED_CODE();
+
+    DbgPrint(TRACE_LEVEL_VERBOSE, ("---> %s res_id=%d\n", __FUNCTION__, m_Id));
+
+    GPU_BOX box;
+    box.x = 0; 
+    box.y = 0;  
+    box.z = 0; 
+    box.width = m_options.width; 
+    box.height = m_options.height; 
+    box.depth = 1;
+        
+    m_adapter->ctrlQueue.SetScanout(scan_id, m_Id, m_options.width, m_options.height, 0, 0);
+    m_adapter->ctrlQueue.ResFlush(m_Id, m_options.width, m_options.height, 0, 0);
+
+    DbgPrint(TRACE_LEVEL_VERBOSE, ("<--- %s res_id=%d\n", __FUNCTION__, m_Id));
+}
+
+#define VIRGL_BIND_RENDER_TARGET (1 << 1)
+#define VIRGL_BIND_SAMPLER_VIEW  (1 << 3)
+#define VIRGL_BIND_DISPLAY_TARGET (1 << 7)
+#define VIRGL_BIND_DISPLAY_TARGET (1 << 7)
+#define VIRGL_BIND_SCANOUT       (1 << 18)
```

Comment:
I think it better to move VIRGL defines to some header to easily add/update it in future

Reply:
This is a part of virgl_hw.h. Looks like we need different parts of this header. We have a similar situation with VirtIO headers, the `/VirtIO/linux/virtio_config.h` is a Linux header file. I suggest placing virgl_hw.h and other required headers in the common/shared folder.

viogpu_allocation.h -> #define VIRGL_RESOURCE_FLAG_MAP_COHERENT

Reply:
Virglrenderer related definitions were moved to `virgl_hw.h`. We can't use `virgl_hw.h` from virglrenderer directly as it uses "non-windowsy" types and we only need a small subset of it, so it's easier to just port things we need similarly to `viogpu.h`